lib/router/src/cache.rs

```diff
@@ -1,426 +1,461 @@
+    pub fn get(&mut self, key: &K) -> Option<V> {
+        if let Some(node) = self.values.get_mut(key) {
+            self.expirations.reset(&node.dq_key, self.expires);
+            return Some(node.value.clone());
         }
-    }
-}
-
-// ===== impl Reserve =====
-
-impl<'a, K: Hash + Eq + 'a, V: 'a, N: Now + 'a> Reserve<'a, K, V, N> {
-    /// Stores a route in the cache.
-    pub fn store(self, key: K, val: V) {
-        let node = Node::new(val.into(), self.now.now());
-        self.vals.insert(key, node);
-    }
-}
-
-// ===== impl Access =====
-
-impl<'a, T: 'a, N: Now + 'a> Deref for Access<'a, T, N> {
-    type Target = T;
-    fn deref(&self) -> &Self::Target {
-        &self.node
-    }
-}
 
-impl<'a, T: 'a, N: Now + 'a> DerefMut for Access<'a, T, N> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        &mut self.node
+        None
     }
-}
 
-impl<'a, T: 'a, N: Now + 'a> Access<'a, T, N> {
-    #[cfg(test)]
-    fn last_access(&self) -> Instant {
-        self.node.last_access
-    }
-}
+    pub fn insert(&mut self, key: K, value: V) -> Option<V> {
+        let node = {
+            let dq_key = self.expirations.insert(key.clone(), self.expires);
+            Node { dq_key, value }
+        };
 
-impl<'a, T: 'a, N: Now + 'a> Drop for Access<'a, T, N> {
-    fn drop(&mut self) {
-        self.node.last_access = self.now.now();
+        self.values.insert(key, node).map(|n| n.value)
     }
-}
 
-// ===== impl Node =====
+    pub fn poll_insert(&mut self) -> Poll<(), Never> {
+        // When checking capacity, only try to remove values if the cache is
+        // at capacity
+        if self.values.len() == self.capacity {
+            match self
+                .expirations
+                .poll()
+                .expect("delay_queue::poll must not fail")
+            {
+                // The cache is at capacity, but we are able to remove a value
+                Async::Ready(Some(expired)) => {
+                    self.values.remove(expired.get_ref());
+                }
+
+                // `Ready(None)` can only be returned when `expirations` is
+                // empty; we know `expirations` is not empty because `values`
+                // is not empty and capacity does not equal zero.
+                Async::Ready(None) => unreachable!("cache expirations cannot be empty"),
+
+                // The cache is at capacity and no values can be removed
+                Async::NotReady => {
+                    return Ok(Async::NotReady);
+                }
+            }
+        }
 
-impl<T> Node<T> {
-    pub fn new(value: T, last_access: Instant) -> Self {
-        Node { value, last_access }
+        Ok(Async::Ready(()))
     }
 
-    pub fn access<'a, N: Now + 'a>(&'a mut self, now: &'a N) -> Access<'a, T, N> {
-        Access { now, node: self }
-    }
+    fn poll_purge(&mut self) -> Poll<(), Error> {
+        while let Some(expired) = try_ready!(self.expirations.poll()) {
```

Comment:
Hmm.. so, I'm having a little trouble trying to understand the semantics of `expirations` -- it seems that it is polled by two tasks, which seems unsafe.

Specifically, I worry that a `poll_insert` can prevent a `poll_purge` from being notified...

Why does `expirations` need to be polled in `poll_insert`?

Reply:
If we choose not to spawn `cache_bg`, then we have the current behavior of no values being removed from the cache. In these cases, `poll_insert` will at some point [branch](https://github.com/linkerd/linkerd2-proxy/pull/247/files/ccb320e266c984268c22eae14b2358e66765f78c#diff-230615b3316b2a089b0d6563c3f74944R109) into the cache being at capacity.

In order to make capacity, we can check if any values should be removed and therefore need to `poll` `expirations`.

This is the reason that there is a lock around `Cache`. The `Future:poll` impl for `PurgeCache` will not `poll_purge` until it has acquired a lock for the cache.